File: GSA/uebung_2/aufgabe_1.py
#! /usr/bin/env python2.7

import sys, math, getopt, numpy

class Loesung:
    
    countC = None
    countA = None
    matrix = None
    
    def __init__(self):
        self.countC = 0 
        self.countA = 0 
        self.matrix = []

    def calls (self):
        return self.countC

    def accesses (self):
        return self.countA 

    def alignsR (self, i, j):
        self.countC +=1
        result = 1
        if i <= 15 and j <= 15:
            if i != 0 and j != 0 :
                result = self.alignsR(i-1, j) + self.alignsR(i, j-1) + self.alignsR(i-1, j-1)
            return result
        return None

    def alignsM (self, n, m):
        if n <= 15 and m <= 15 :
            for i in range (0, n + 1) :
                self.matrix.append([])
                for j in range (0, m + 1) : 
                    if i == 0 or j == 0 :
                        self.matrix[i].append(1)
                    else :
                        self.countA += 1
                        self.matrix[i].append(self.matrix[i-1][j] + self.matrix[i][j-1] + self.matrix[i-1][j-1])
            return self.matrix[n][m]
        return None


def main(argv):

    arg1 = None
    arg2 = None

    if len(argv) == 2 :
        
        arg1 = int(argv[0])
        arg2 = int(argv[1])
        
        loesung = Loesung()
        
        resultM = loesung.alignsM(arg1, arg2)
        accesses = loesung.accesses()
        
        resultR = loesung.alignsR(arg1, arg2)
        calls = loesung.calls()
                
        if resultR != None and resultM != None:
            print arg1, arg2, resultM, "(" + str(accesses) + " accesses)"
            print arg1, arg2, resultR, "(" + str(calls) + " calls)"
            return(0)


    print("Falshe eingabe!")
    if arg1 == None or arg2 == None :
        print("Bitte, geben Sie zwei Zahlen ein!")
        return(0)
    print("i="+str(arg1)+" j="+str(arg2))
    return(0)



if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
    

File: GSA/uebung_2/aufgabe_1.c
#include <stdio.h>
#include <malloc.h>

long int alignsR (int var1, int var2, int *counter)
{
	*counter = *counter +1;
	if(var1 == 0 || var2 == 0) {
		return 1;
	}
	return alignsR(var1-1, var2, counter) + alignsR(var1, var2-1, counter) + alignsR(var1-1, var2-1, counter);
}

long int alignsI (int var1, int var2, int *counter)
{
	int i, j;
	int n, m;
	int x;

	n = var1 + 1;
	m = var2 + 1;

	int matrix[16][16];

	for(i = 0; i < n; i++) {
		for(j = 0; j < m; j++) {
			if( i == 0 || j == 0) {
				matrix[i][j] = 1;
			} else {
				*counter = *counter + 3;
				matrix[i][j] = matrix[i-1][j] + matrix[i][j-1] + matrix[i-1][j-1];
			}
		}
	}

	return matrix[var1][var2];
}


int main (int argc, char *argv[])
{
	int counterR = 0, counterI = 0;

	int var1, var2, result;

	if(argc != 3) {
		printf("Falshe Eingabe!");
		return(0);
	}

	var1 = atoi(argv[1]);
	var2 = atoi(argv[2]);

	if(var1 > 15 || var2 > 15) {
		printf("Falshe Eingabe! x <= 15 !");
		return(0);
	}

	printf("n: %d, m: %d Result: %d ", var1, var2, alignsI(var1, var2, &counterI));
	printf("(%d accesses)\n", counterI);

	printf("n: %d, m: %d Result: %d ", var1, var2, alignsR(var1, var2, &counterR));
	printf("(%d calls)\n", counterR);

	return(0);
}
